platform/pulpcore/download/http.py

```diff
@@ -118,13 +120,25 @@ def session(self):
             The session can be shared between download but this needs to be
             facilitated by a 3rd object by setting the `context` to be the same.
         """
-        with self.context as context:
-            try:
-                return context.session
-            except AttributeError:
-                session = Session()
-                context.session = session
-                return session
+        parsed_url = urlparse(self.url)
+        properties = (
+            Session,
+            parsed_url.hostname,
+            parsed_url.port,
+            self.user.name,
+            self.user.password,
+            self.proxy_url,
+            self.ssl.client_certificate
+        )
+        key = hash(properties)
```

Comment:
The real protection from race conditions is provided by the `Context` context manager which acquires the mutex for the get-or-create logic.  When porting this I forgot the `with self.context as context:`.  I added it back and pushed the change.

As for eliminating the `SessionCache`, it seemed more appropriate that the user of the cache be responsible for constructing the key and doing the get or create/add logic.  Just like we don't subclass a `dict ` and add methods to building the key and get_or_create().